Rename package and redesign types (#5)
JuliaImages/FluorophoreColors.jl: docs/src/index.md

```@meta
CurrentModule = MultiChannelColors
```

# MultiChannelColors

[MultiChannelColors](https://github.com/JuliaImages/MultiChannelColors.jl) aims to support "unconventional colors," such as might arise in applications like multichannel fluorescence microscopy and hyperspectral imaging. Consistent with the philosophy of the [JuliaImages ecosystem](https://juliaimages.org/latest/), this package allows you to bundle together the different color channels into a "color object," and many color objects can be stored in an array. Having each entry of the array represent a complete pixel or voxel makes it much easier to write generic code supporting a wide range of image types.

## Installation

Install the package with `add MultiChannelColors` from the `pkg>` prompt, which you access by typing `]` from the `julia>` prompt. See the [Pkg documentation](https://pkgdocs.julialang.org/v1/getting-started/) for more information.

## Usage

Use the package interactively or in code with

> ```jldoctest demo

Reviewer: Just a confirmation: is the `jldoctest` deliberately used here rather than the [`@repl` block](https://juliadocs.github.io/Documenter.jl/stable/man/syntax/#@repl-block)?

Author: Yes, though there are many ways to solve the same problem. Basically I want doctests to make sure the code keeps working, and I don't want to have to use `setup=:(using MultiChannelColors)` on each doctest. So I created a named doctest, `demo`, and am loading the package in the first instance.

julia> using MultiChannelColors
```

In addition to giving access to specific types defined below, this will import the namespaces of [FixedPointNumbers](https://github.com/JuliaMath/FixedPointNumbers.jl) (which harmonizes the interpretation of "integer" and "floating-point" pixel-encodings) and [ColorTypes](https://github.com/JuliaGraphics/ColorTypes.jl) (which defines core color types and low-level manipulation). It will also define arithmetic for colors such as RGB (see [ColorVectorSpace](https://github.com/JuliaGraphics/ColorVectorSpace.jl)).

The color types in this package support two fundamental categories of operations:

> - arithmetic operations such as `+` and `-` and multiplying or dividing by a scalar. You can also scale each color channel independently with `⊙` (obtained with `\odotTAB`) or its synonym `hadamard`, e.g., `g ⊙ c` where `c` is a color object defined in this package and `g` is a tuple of real numbers (the "gains").

Reviewer: As the `?⊙` gives:

```suggestion
- arithmetic operations such as `+` and `-` and multiplying or dividing by a scalar. You can also scale each color channel independently with `⊙` (obtained with `\odot<tab>`) or its synonym `hadamard`, e.g., `g ⊙ c` where `c` is a color object defined in this package and `g` is a tuple of real numbers (the "gains").
```

> - extracting the independent channel intensities as a tuple with `Tuple(c)`.

Reviewer: TIL 😆

When creating `c`, you have two choices which primarily affect visualization:

> - to use "bare" colors that store the multichannel data but lack any default conversion to other color spaces. This might be most appropriate if you have more than 3 channels, for which there may be many different ways to visualize the data they encode.

Reviewer: ```suggestion
- to use ["bare" colors](@ref index_multichannelcolor) that store the multichannel data but lack any default conversion to other color spaces. This might be most appropriate if you have more than 3 channels, for which there may be many different ways to visualize the data they encode.
```

> - to use colors with built-in conversion to RGB, making them work automatically in standard visualization tools. This may be most appropriate when you have 3 or fewer channels.

Reviewer: ```suggestion
- to use [colors with built-in conversion to RGB](@ref index_colormixture), making them work automatically in standard visualization tools. This may be most appropriate when you have 3 or fewer channels.
```

Both options will be discussed below. See the [JuliaImages documentation on visualization](https://juliaimages.org/latest/install/#sec_visualization) for more information about tools for viewing images.

> ### "Bare" colors: `MultiChannelColor`

Reviewer: ```suggestion
### ["Bare" colors: `MultiChannelColor`](@id index_multichannelcolor)
```

The usual way to visualize such an object is to define a custom function that converts such colors to more conventional colors (`RGB` or `Gray`). For example, we might compute the [Enhanced Vegetation Index](https://www.usgs.gov/landsat-missions/landsat-enhanced-vegetation-index)
and render positive values in green and negative values in magenta:

```jldoctest demo
julia> function evi(c::MultiChannelColor{T,11}) where T<:FixedPoint
           # Valid for Landsat 8 with 11 spectral bands
           b = Tuple(c)                      # extract the bands
           evi = 2.5f0 * (b[5] - b[4]) / (b[5] + 6*b[4] - 7.5f0*b[2] + eps(T))
           return evi > 0 ? RGB(0, evi, 0) : RGB(-evi, 0, -evi)
       end;

julia> evi(c)
RGB{Float32}(0.0f0,0.17894554f0,0.0f0)
```

If `img` is a whole image of such pixels, `evi.(img)` converts the entire array to RGB. For large data, you might prefer to use the [MappedArrays package](https://github.com/JuliaArrays/MappedArrays.jl) to do such conversions "lazily" (on an as-needed basis) to avoid exhausting computer memory:

```julia
julia> using MappedArrays

julia> imgrgb = mappedarray(evi, img);
```

> ### RGB-convertible colors: `ColorMixture`

Reviewer: ```suggestion
### [RGB-convertible colors: `ColorMixture`](@id index_colormixture)
```

``

Depending on the the `rgbⱼ` and `iⱼ`, values may exceed the 0-to-1 colorscale of RGBs.
Conversion to `RGB{Float32}` may be safer than `RGB{T}` where `T` is limited to 0-to-1.
It is also faster, as the result does not have to be checked for whether it exceeds the bounds of the type.
(To prevent overflow, all internal operations are performed using floating-point intermediates even if you want a `FixedPoint` output.)

!!! note
    While `ColorMixture` objects can be converted to RGB, they are *not* AbstractRGB
    colors: `red(c)`, `green(c)`, and `blue(c)` are not defined for `c::ColorMixture`, and low-level utilities
    like `mapc` operate on the raw channel intensities rather than the RGB values.

There are several ways you can create these colors. An easy approach is to define the type through a "template" object:

```jldoctest demo
julia> ctemplate = ColorMixture{Float32}((RGB(0,1,0), RGB(1,0,0)))
(0.0₁, 0.0₂)
```

`ctemplate` is an all-zeros `ColorMixture` object, but can be used to construct arbitrary `c` with specified intensities:

```jldoctest demo
julia> typeof(ctemplate)
ColorMixture{Float32, 2, (RGB{N0f8}(0.0,1.0,0.0), RGB{N0f8}(1.0,0.0,0.0))}

> julia> c = ctemplate(0.2, 0.4)

Reviewer: neat!

    Julia 1.7 and higher can use "aggressive" constant propagation to solve inference problems that may reduce performance on Julia 1.6.

### Importing external data

When objects are not created by code but instead loaded from an external source such as a file, you have several avenues for creating arrays of multichannel color objects. There are two particularly common cases:

1. If the imported data are an array `A` of size `(nc, m, n)`, where `nc` is the number of color channels (i.e., color is the fastest dimension), then use `reinterpret(reshape, C, A)` where `C` is the color type you want to use (e.g., `MultiChannelColor{T,nc}` or `ColorMixture{T,nc,rgbs}`). For instance, Landsat 8 data might look something like this:

   ```julia
   A = rand(0x0000:0x0fff, 11, 100, 100);
   img = reinterpret(reshape, MultiChannelColor{N4f12,11}, A);
   ```

2. If the imported data have the color channel last, or use separate arrays for each channel, use the [StructArrays package](https://github.com/JuliaArrays/StructArrays.jl). For example:

   ```julia
   A = rand(0x0000:0x0fff, 100, 100, 11);
   img = StructArray{MultiChannelColor{N4f12,11}}(A; dims=3)

>    ```

Reviewer: If we provide a better tool in ImageCore in the spirit of https://github.com/JuliaImages/ImageCore.jl/pull/179, then this story might be changed.
Might be worth leaving a note about future improvement.

---

I was thinking of something like:

```julia
module ImageLayouts
const CHW = DenseImageLayout{:CHW}()
const CWH = DenseImageLayout{:CWH}()
const HWC = DenseImageLayout{:HWC}()
end
```

so that users can use a unified interface:

```julia
using ImageCore # which exports the ImageLayouts module

colorview(ImageLayouts.CHW, img)
```

we might end up with a new function, though -- pretty much like the `strel` concept I introduced in ImageMorphology.

Author: Yes, I really like the idea of these custom layouts!

## Additional features

### Fluorophores

This package also exports a lookup table for common [fluorophores](https://en.wikipedia.org/wiki/Fluorophore). If desired, these can be used as the `rgbⱼ` values for `ColorMixture` channels. For example:

```jldoctest demo
julia> channels = (fluorophore_rgb["EGFP"], fluorophore_rgb["tdTomato"])
(RGB{N0f8}(0.0,0.925,0.365), RGB{N0f8}(1.0,0.859,0.0))

julia> ctemplate = ColorMixture{N0f16}(channels)
(0.0N0f16₁, 0.0N0f16₂)
```

If you'll be hard-coding the name of the fluorophore, consider using a slightly different syntax:

```jldoctest demo

> julia> channels = (fluorophore_rgb"EGFP", fluorophore_rgb"tdTomato")

Reviewer: Will `fluorophore"EGFP"` be easier to use? I'm not sure. -- just that I have never seen underscores used in string macros.

Author: No non-RGB fluorophore currently. I guess I would think that `fluorophore"EGFP"` might return an entire object that is a complete characterization of the fluorophore, e.g., full spectra, cross section, quantum yield, and so on ([this list](https://en.wikipedia.org/wiki/Fluorophore#Fluorescent_proteins) only scratches the surface). In contrast, `fluorophore_rgb` makes it clear that what you're getting is an RGB value that represents the fluorophore.
